Panics when appending to "empty"

**What steps did you take:**

``` yaml
#! dreams.yaml
#@ load("@ytt:overlay", "overlay")

#@overlay/match by=overlay.subset({"name": "Sandman", "profession": "King of Dreams"})
#@overlay/append
---
name: Matthew
profession: Raven
```

```shell
ytt -f dreams.yaml
```

**What happened:**

```
panic: runtime error: index out of range [-1]

goroutine 1 [running]:
github.com/vmware-tanzu/carvel-ytt/pkg/yttlibrary/overlay.Op.appendDocument(...)
        github.com/vmware-tanzu/carvel-ytt/pkg/yttlibrary/overlay/document.go:197
github.com/vmware-tanzu/carvel-ytt/pkg/yttlibrary/overlay.Op.applyDocSet({{0x148b260, 0xc0000aed08}, {0x1516700, 0xc0001595e0}, 0xc0001abd60, 0x0}, {0x19d9090, 0x1, 0x0}, 0xc0001595e0, ...)
        github.com/vmware-tanzu/carvel-ytt/pkg/yttlibrary/overlay/op.go:149 +0x805
github.com/vmware-tanzu/carvel-ytt/pkg/yttlibrary/overlay.Op.apply({{0x148b260, 0xc0000aed08}, {0x1516700, 0xc0001595e0}, 0xc0001abd60, 0x0}, {0x148b260, 0xc0000aed20}, {0x1516700, 0xc0001595e0}, ...)
        github.com/vmware-tanzu/carvel-ytt/pkg/yttlibrary/overlay/op.go:55 +0x35c
github.com/vmware-tanzu/carvel-ytt/pkg/yttlibrary/overlay.Op.Apply({{0x148b260, 0xc0000aed08}, {0x1516700, 0xc0001595e0}, 0xc0001abd60, 0x0})
        github.com/vmware-tanzu/carvel-ytt/pkg/yttlibrary/overlay/op.go:27 +0x14f
github.com/vmware-tanzu/carvel-ytt/pkg/workspace.OverlayPostProcessing.Apply({0xc0001b3a58})
        github.com/vmware-tanzu/carvel-ytt/pkg/workspace/overlay_post_processing.go:63 +0x716
github.com/vmware-tanzu/carvel-ytt/pkg/workspace.(*LibraryExecution).Eval(0xc0001b3a58, 0x0, {0x0, 0x0, 0xc0000ae5e8}, {0x0, 0xc0000a1960, 0x0})
        github.com/vmware-tanzu/carvel-ytt/pkg/workspace/library_execution.go:183 +0x7b
github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template.(*Options).RunWithFiles(0xc0000db540, {{0xc0000b6100, 0x2, 0xc}}, {0x1689890, 0xc0001532f0})
        github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template/cmd.go:164 +0x570
github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template.(*Options).Run(0xc0000db540)
        github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template/cmd.go:100 +0x4b7
github.com/vmware-tanzu/carvel-ytt/pkg/cmd.NewCmd.func1(0x0, {0xc0000ce9c0, 0x0, 0x0})
        github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template.go:22 +0x1d
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0x0, {0xc0000ce9c0, 0x0, 0x2})
        github.com/cppforlife/cobrautil@v0.0.0-20200514214827-bb86e6965d72/misc.go:45 +0x6d
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc00013ea00, {0xc0000ce9c0, 0x0, 0x2})
        github.com/cppforlife/cobrautil@v0.0.0-20200514214827-bb86e6965d72/misc.go:45 +0x6d
github.com/spf13/cobra.(*Command).execute(0xc00013ea00, {0xc0000a2190, 0x2, 0x2})
        github.com/spf13/cobra@v1.5.0/command.go:872 +0x624
github.com/spf13/cobra.(*Command).ExecuteC(0xc00013ea00)
        github.com/spf13/cobra@v1.5.0/command.go:990 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.5.0/command.go:918
main.main()
        github.com/vmware-tanzu/carvel-ytt/cmd/ytt/ytt.go:21 +0xd6
```

**What did you expect:**

A helpful error message.


**Anything else you would like to add:**
[Additional information that will assist in solving the issue.]


**Environment:**

- ytt version (use `ytt --version`): `ytt version 0.42.0`
- OS (e.g. from `/etc/os-release`): macOS

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

While trying to write a template test for this, I realised that I can't reproduce it with a `*.tpltest`:

```yaml
#! pkg/yamltemplate/filetests/ytt-library/overlay/append-to-nothing.tpltest

#@ load("@ytt:overlay", "overlay")

#@overlay/match by=overlay.subset({"name": "Sandman", "profession": "King of Dreams"})
#@overlay/append
---
name: Matthew
profession: Raven

+++

---
```
```
--- FAIL: TestYAMLTemplate (0.10s)
    --- FAIL: TestYAMLTemplate/filetests/ytt-library/overlay/append-to-nothing.tpltest (0.00s)
        filetests.go:131: not equal

            ### result 32 chars:
            >>>name: Matthew
            profession: Raven
            <<<
            ###expected 4 chars:
            >>>---
            <<<
```
First I thought it maybe already fixed on `develop`, but that's not the case:
```
❯ go run cmd/ytt/ytt.go -f dreams.yaml
panic: runtime error: index out of range [-1]

...
```

So I think the template tests don't fully replicate `ytt` command's entry point.

Either way, from what I can tell, [this](https://github.com/vmware-tanzu/carvel-ytt/blob/develop/pkg/yttlibrary/overlay/document.go#L197) is the offending line as per trace and studying the code.

I think it would make sense to append _nothing_ if there are no left-hand docs. The same probably goes for arrays.

--- a/pkg/yamltemplate/filetests/ytt-library/overlay/insert-into-array-via.tpltest
+++ b/pkg/yamltemplate/filetests/ytt-library/overlay/insert-into-array-via.tpltest
@@ -1,0 +1,72 @@
+#@ load("@ytt:overlay", "overlay")
+
+---
+#@ def test1_left():
+- item: 1
+- item: 2
+- item: 3
+#@ end
+
+#@ def is_item(indexOrKey, left, right):
+#@   return "item" in left
+#@ end
+
+#@ def insert_function(left, right):
+#@   return {"left": left, "right": right, "combined": {"item": left["item"] + right["item"]}}
+#@ end
+
+---
+#@ def test1_right():
+#@overlay/match by=is_item, expects=3
+#@overlay/insert before=True, via=insert_function
+- item: 100
+#@overlay/match by=is_item, expects=3
+#@overlay/insert after=True, via=insert_function
+- item: 200
+#@ end
+
+---
+test1: #@ overlay.apply(test1_left(), test1_right())
+
++++
+
+test1:
+- left:
+    item: 1
+  right:
+    item: 100
+  combined:
+    item: 101
+- item: 1
+- left:
+    item: 1
+  right:
+    item: 200
+  combined:
+    item: 201
+- left:
+    item: 2
+  right:
+    item: 100
+  combined:
+    item: 102
+- item: 2
+- left:
+    item: 2
+  right:
+    item: 200
+  combined:
+    item: 202
+- left:
+    item: 3
+  right:
+    item: 100
+  combined:
+    item: 103
+- item: 3
+- left:
+    item: 3
+  right:
+    item: 200
+  combined:
+    item: 203

--- a/pkg/yamltemplate/filetests/ytt-library/overlay/insert-into-doc-via.tpltest
+++ b/pkg/yamltemplate/filetests/ytt-library/overlay/insert-into-doc-via.tpltest
@@ -1,0 +1,82 @@
+#@ load("@ytt:template", "template")
+#@ load("@ytt:overlay", "overlay")
+
+#@ def test1_left():
+---
+item: 1
+---
+item: 2
+---
+item: 3
+#@ end
+
+#@ def is_item(indexOrKey, left, right):
+#@   return "item" in left
+#@ end
+
+#@ def insert_function(left, right):
+#@   return {"left": left, "right": right, "combined": {"item": left["item"] + right["item"]}}
+#@ end
+
+#@ def test1_right():
+#@overlay/match by=is_item, expects=3
+#@overlay/insert before=True, via=insert_function
+---
+item: 100
+#@overlay/match by=is_item, expects=3
+#@overlay/insert after=True, via=insert_function
+---
+item: 200
+#@ end
+
+--- #@ template.replace(overlay.apply(test1_left(), test1_right()))
+
++++
+
+left:
+  item: 1
+right:
+  item: 100
+combined:
+  item: 101
+---
+item: 1
+---
+left:
+  item: 1
+right:
+  item: 200
+combined:
+  item: 201
+---
+left:
+  item: 2
+right:
+  item: 100
+combined:
+  item: 102
+---
+item: 2
+---
+left:
+  item: 2
+right:
+  item: 200
+combined:
+  item: 202
+---
+left:
+  item: 3
+right:
+  item: 100
+combined:
+  item: 103
+---
+item: 3
+---
+left:
+  item: 3
+right:
+  item: 200
+combined:
+  item: 203

--- a/pkg/yamltemplate/filetests/ytt-library/overlay/insert-into-via-err.tpltest
+++ b/pkg/yamltemplate/filetests/ytt-library/overlay/insert-into-via-err.tpltest
@@ -1,0 +1,22 @@
+#@ load("@ytt:template", "template")
+#@ load("@ytt:overlay", "overlay")
+
+#@ def test1_left():
+---
+item: 1
+#@ end
+
+#@ def test1_right():
+#@overlay/match by=overlay.all, expects=1
+#@overlay/insert before=True, via="not a function"
+---
+#@ end
+
+--- #@ template.replace(overlay.apply(test1_left(), test1_right()))
+
++++
+
+ERR:
+- overlay.apply: Document on line stdin:12: Expected 'overlay/insert' annotation keyword argument 'via' to be function, but was starlark.String
+    in <toplevel>
+      stdin:15 | --- #@ template.replace(overlay.apply(test1_left(), test1_right()))

--- a/pkg/yamltemplate/filetests/ytt-library/overlay/replace-func-err.tpltest
+++ b/pkg/yamltemplate/filetests/ytt-library/overlay/replace-func-err.tpltest
@@ -1,0 +1,22 @@
+#@ load("@ytt:template", "template")
+#@ load("@ytt:overlay", "overlay")
+
+#@ def test1_left():
+---
+item: 1
+#@ end
+
+#@ def test1_right():
+#@overlay/match by=overlay.all, expects=1
+#@overlay/replace via="not a function"
+---
+#@ end
+
+--- #@ template.replace(overlay.apply(test1_left(), test1_right()))
+
++++
+
+ERR:
+- overlay.apply: Document on line stdin:12: Expected 'overlay/replace' annotation keyword argument 'via' to be function, but was starlark.String
+    in <toplevel>
+      stdin:15 | --- #@ template.replace(overlay.apply(test1_left(), test1_right()))

--- a/pkg/yttlibrary/overlay/array.go
+++ b/pkg/yttlibrary/overlay/array.go
@@ -159,7 +159,7 @@ func (o Op) insertArrayItem(
 		return err
 	}
 
-	insertAnn, err := NewInsertAnnotation(newItem)
+	insertAnn, err := NewInsertAnnotation(newItem, o.Thread)
 	if err != nil {
 		return err
 	}
@@ -171,12 +171,23 @@ func (o Op) insertArrayItem(
 		for _, leftIdx := range leftIdxs {
 			if i == leftIdx {
 				matched = true
+
+				newVal, err := insertAnn.Value(leftItem)
+				if err != nil {
+					return err
+				}
+				insertItem := newItem.DeepCopy()
+				err = insertItem.SetValue(newVal)
+				if err != nil {
+					return err
+				}
+
 				if insertAnn.IsBefore() {
-					updatedItems = append(updatedItems, newItem.DeepCopy())
+					updatedItems = append(updatedItems, insertItem)
 				}
 				updatedItems = append(updatedItems, leftItem)
 				if insertAnn.IsAfter() {
-					updatedItems = append(updatedItems, newItem.DeepCopy())
+					updatedItems = append(updatedItems, insertItem)
 				}
 				break
 			}

--- a/pkg/yttlibrary/overlay/document.go
+++ b/pkg/yttlibrary/overlay/document.go
@@ -156,7 +156,7 @@ func (o Op) insertDocument(
 		return err
 	}
 
-	insertAnn, err := NewInsertAnnotation(newDoc)
+	insertAnn, err := NewInsertAnnotation(newDoc, o.Thread)
 	if err != nil {
 		return err
 	}
@@ -169,12 +169,23 @@ func (o Op) insertDocument(
 			for _, leftIdx := range leftIdxs {
 				if leftIdx[0] == i && leftIdx[1] == j {
 					matched = true
+
+					newVal, err := insertAnn.Value(leftItem)
+					if err != nil {
+						return err
+					}
+					insertDoc := newDoc.DeepCopy()
+					err = insertDoc.SetValue(newVal)
+					if err != nil {
+						return err
+					}
+
 					if insertAnn.IsBefore() {
-						updatedDocs = append(updatedDocs, newDoc.DeepCopy())
+						updatedDocs = append(updatedDocs, insertDoc)
 					}
 					updatedDocs = append(updatedDocs, leftItem)
 					if insertAnn.IsAfter() {
-						updatedDocs = append(updatedDocs, newDoc.DeepCopy())
+						updatedDocs = append(updatedDocs, insertDoc)
 					}
 					break
 				}

--- a/pkg/yttlibrary/overlay/insert_annotation.go
+++ b/pkg/yttlibrary/overlay/insert_annotation.go
@@ -22,7 +22,7 @@ type InsertAnnotation struct {
 	newItem template.EvaluationNode
 	before  bool
 	after   bool
-	via     *starlark.Value
+	via     *starlark.Callable
 	thread  *starlark.Thread
 }
 
@@ -64,7 +64,13 @@ func NewInsertAnnotation(newItem template.EvaluationNode, thread *starlark.Threa
 			annotation.after = resultBool
 
 		case InsertAnnotationKwargVia:
-			annotation.via = &kwarg[1]
+			via, ok := kwarg[1].(starlark.Callable)
+			if !ok {
+				return InsertAnnotation{}, fmt.Errorf(
+					"Expected '%s' annotation keyword argument '%s' to be function, "+
+						"but was %T", AnnotationInsert, kwargName, kwarg[1])
+			}
+			annotation.via = &via
 
 		default:
 			return annotation, fmt.Errorf(
@@ -86,29 +92,22 @@ func (a InsertAnnotation) Value(existingNode template.EvaluationNode) (interface
 		return newNode.GetValues()[0], nil
 	}
 
-	switch typedVal := (*a.via).(type) {
-	case starlark.Callable:
-		var existingVal interface{}
-		if existingNode != nil {
-			existingVal = existingNode.DeepCopyAsInterface().(template.EvaluationNode).GetValues()[0]
-		} else {
-			existingVal = nil
-		}
-
-		viaArgs := starlark.Tuple{
-			yamltemplate.NewGoValueWithYAML(existingVal).AsStarlarkValue(),
-			yamltemplate.NewGoValueWithYAML(newNode.GetValues()[0]).AsStarlarkValue(),
-		}
-
-		result, err := starlark.Call(a.thread, *a.via, viaArgs, []starlark.Tuple{})
-		if err != nil {
-			return nil, err
-		}
-
-		return tplcore.NewStarlarkValue(result).AsGoValue()
-
-	default:
-		return nil, fmt.Errorf("Expected '%s' annotation keyword argument 'via'"+
-			" to be function, but was %T", AnnotationInsert, typedVal)
+	var existingVal interface{}
+	if existingNode != nil {
+		existingVal = existingNode.DeepCopyAsInterface().(template.EvaluationNode).GetValues()[0]
+	} else {
+		existingVal = nil
 	}
+
+	viaArgs := starlark.Tuple{
+		yamltemplate.NewGoValueWithYAML(existingVal).AsStarlarkValue(),
+		yamltemplate.NewGoValueWithYAML(newNode.GetValues()[0]).AsStarlarkValue(),
+	}
+
+	result, err := starlark.Call(a.thread, *a.via, viaArgs, []starlark.Tuple{})
+	if err != nil {
+		return nil, err
+	}
+
+	return tplcore.NewStarlarkValue(result).AsGoValue()
 }

--- a/pkg/yttlibrary/overlay/insert_annotation.go
+++ b/pkg/yttlibrary/overlay/insert_annotation.go
@@ -5,20 +5,33 @@ package overlay
 
 import (
 	"fmt"
-
 	"github.com/k14s/starlark-go/starlark"
 	"github.com/vmware-tanzu/carvel-ytt/pkg/template"
 	tplcore "github.com/vmware-tanzu/carvel-ytt/pkg/template/core"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/yamltemplate"
+)
+
+// Kwargs of overlay/insert
+const (
+	InsertAnnotationKwargBefore string = "before"
+	InsertAnnotationKwargAfter  string = "after"
+	InsertAnnotationKwargVia    string = "via"
 )
 
 type InsertAnnotation struct {
 	newItem template.EvaluationNode
 	before  bool
 	after   bool
+	via     *starlark.Value
+	thread  *starlark.Thread
 }
 
-func NewInsertAnnotation(newItem template.EvaluationNode) (InsertAnnotation, error) {
-	annotation := InsertAnnotation{newItem: newItem}
+// NewInsertAnnotation returns a new InsertAnnotation for the given node and with the given Starlark thread
+func NewInsertAnnotation(newItem template.EvaluationNode, thread *starlark.Thread) (InsertAnnotation, error) {
+	annotation := InsertAnnotation{
+		newItem: newItem,
+		thread:  thread,
+	}
 	anns := template.NewAnnotations(newItem)
 
 	if !anns.Has(AnnotationInsert) {
@@ -36,19 +49,22 @@ func NewInsertAnnotation(newItem template.EvaluationNode) (InsertAnnotation, err
 		kwargName := string(kwarg[0].(starlark.String))
 
 		switch kwargName {
-		case "before":
+		case InsertAnnotationKwargBefore:
 			resultBool, err := tplcore.NewStarlarkValue(kwarg[1]).AsBool()
 			if err != nil {
 				return InsertAnnotation{}, err
 			}
 			annotation.before = resultBool
 
-		case "after":
+		case InsertAnnotationKwargAfter:
 			resultBool, err := tplcore.NewStarlarkValue(kwarg[1]).AsBool()
 			if err != nil {
 				return InsertAnnotation{}, err
 			}
 			annotation.after = resultBool
+
+		case InsertAnnotationKwargVia:
+			annotation.via = &kwarg[1]
 
 		default:
 			return annotation, fmt.Errorf(
@@ -61,3 +77,38 @@ func NewInsertAnnotation(newItem template.EvaluationNode) (InsertAnnotation, err
 
 func (a InsertAnnotation) IsBefore() bool { return a.before }
 func (a InsertAnnotation) IsAfter() bool  { return a.after }
+
+// Value returns the new value for the given, existing node. If `via` is not provided, the value of the existing
+// node is returned, otherwise the result of `via`.
+func (a InsertAnnotation) Value(existingNode template.EvaluationNode) (interface{}, error) {
+	newNode := a.newItem.DeepCopyAsInterface().(template.EvaluationNode)
+	if a.via == nil {
+		return newNode.GetValues()[0], nil
+	}
+
+	switch typedVal := (*a.via).(type) {
+	case starlark.Callable:
+		var existingVal interface{}
+		if existingNode != nil {
+			existingVal = existingNode.DeepCopyAsInterface().(template.EvaluationNode).GetValues()[0]
+		} else {
+			existingVal = nil
+		}
+
+		viaArgs := starlark.Tuple{
+			yamltemplate.NewGoValueWithYAML(existingVal).AsStarlarkValue(),
+			yamltemplate.NewGoValueWithYAML(newNode.GetValues()[0]).AsStarlarkValue(),
+		}
+
+		result, err := starlark.Call(a.thread, *a.via, viaArgs, []starlark.Tuple{})
+		if err != nil {
+			return nil, err
+		}
+
+		return tplcore.NewStarlarkValue(result).AsGoValue()
+
+	default:
+		return nil, fmt.Errorf("Expected '%s' annotation keyword argument 'via'"+
+			" to be function, but was %T", AnnotationInsert, typedVal)
+	}
+}

--- a/pkg/yttlibrary/overlay/replace_annotation.go
+++ b/pkg/yttlibrary/overlay/replace_annotation.go
@@ -20,7 +20,7 @@ const (
 type ReplaceAnnotation struct {
 	newNode template.EvaluationNode
 	thread  *starlark.Thread
-	via     *starlark.Value
+	via     *starlark.Callable
 	orAdd   bool
 }
 
@@ -36,7 +36,13 @@ func NewReplaceAnnotation(newNode template.EvaluationNode, thread *starlark.Thre
 
 		switch kwargName {
 		case ReplaceAnnotationKwargVia:
-			annotation.via = &kwarg[1]
+			via, ok := kwarg[1].(starlark.Callable)
+			if !ok {
+				return ReplaceAnnotation{}, fmt.Errorf(
+					"Expected '%s' annotation keyword argument '%s' to be function, "+
+						"but was %T", AnnotationReplace, kwargName, kwarg[1])
+			}
+			annotation.via = &via
 
 		case ReplaceAnnotationKwargOrAdd:
 			resultBool, err := tplcore.NewStarlarkValue(kwarg[1]).AsBool()
@@ -63,33 +69,26 @@ func (a ReplaceAnnotation) Value(existingNode template.EvaluationNode) (interfac
 		return newNode.GetValues()[0], nil
 	}
 
-	switch typedVal := (*a.via).(type) {
-	case starlark.Callable:
-		var existingVal interface{}
-		if existingNode != nil {
-			// Make sure original nodes are not affected in any way
-			existingVal = existingNode.DeepCopyAsInterface().(template.EvaluationNode).GetValues()[0]
-		} else {
-			existingVal = nil
-		}
-
-		viaArgs := starlark.Tuple{
-			yamltemplate.NewGoValueWithYAML(existingVal).AsStarlarkValue(),
-			yamltemplate.NewGoValueWithYAML(newNode.GetValues()[0]).AsStarlarkValue(),
-		}
-
-		// TODO check thread correctness
-		result, err := starlark.Call(a.thread, *a.via, viaArgs, []starlark.Tuple{})
-		if err != nil {
-			return nil, err
-		}
-
-		return tplcore.NewStarlarkValue(result).AsGoValue()
-
-	default:
-		return nil, fmt.Errorf("Expected '%s' annotation keyword argument 'via'"+
-			" to be function, but was %T", AnnotationReplace, typedVal)
+	var existingVal interface{}
+	if existingNode != nil {
+		// Make sure original nodes are not affected in any way
+		existingVal = existingNode.DeepCopyAsInterface().(template.EvaluationNode).GetValues()[0]
+	} else {
+		existingVal = nil
 	}
+
+	viaArgs := starlark.Tuple{
+		yamltemplate.NewGoValueWithYAML(existingVal).AsStarlarkValue(),
+		yamltemplate.NewGoValueWithYAML(newNode.GetValues()[0]).AsStarlarkValue(),
+	}
+
+	// TODO check thread correctness
+	result, err := starlark.Call(a.thread, *a.via, viaArgs, []starlark.Tuple{})
+	if err != nil {
+		return nil, err
+	}
+
+	return tplcore.NewStarlarkValue(result).AsGoValue()
 }
 
 func (a ReplaceAnnotation) OrAdd() bool { return a.orAdd }

